### PR TITLE
add more "normalmaps_smooth"s

### PR DESCRIPTION
**Pull request by HybridDog**
this allows generating 16x16 normalmaps
and it changes the use of the normalmaps_smooth setting a bit

screnshots are there https://forum.minetest.net/viewtopic.php?p=168369#p168369

### CODE REVIEW

**Comment by HybridDog** [comment]
the steel block:
steelbmpmp

File: src/shader.cpp [hunk 1]
```diff
@@ -689,14 +689,35 @@ ShaderInfo generate_shader(std::string name, u8 material_type, u8 drawtype,
...
+			sample_step = 0.0078125; // 1/128
+			break;
+		case 8:
+			sample_step = 0.00390625; // 1/256
+			break;
+		case 9:
+			sample_step = 0.001953125; // 1/512
+			break;
+		case 10:
+			sample_step = 0.0009765625; // 1/1024
```
Comment (ShadowNinja) [inline src/shader.cpp]: This should use something more general like `sample_step = 1 / pow(2, smooth);` (with some sanity checks of course).

**Comment by HybridDog** [comment]
ok, l saw that the steelblock screenshot used a _normal.png…

**Comment by HybridDog** [comment]
>This should use something more general like sample_step = 1 / pow(2, smooth); (with some sanity checks of course).

done, but l don't know what you mean with sanity checks

File: src/shader.cpp [hunk 2]
```diff
@@ -686,24 +686,8 @@ ShaderInfo generate_shader(std::string name, u8 material_type, u8 drawtype,
...
-		case 2:
-			sample_step = 0.001953125; // 1.0 / 512.0
-			break;
-		default:
-			sample_step = 0.0078125;
-			break;
-		}
 		shaders_header += "#define SAMPLE_STEP ";
-		shaders_header += ftos(sample_step);
+		shaders_header += ftos(1 / pow(2, (u8)g_settings->getFloat("normalmaps_smooth")));
```
Comment (Zeno-) [inline src/shader.cpp]: Let's take the case of `g_settings->getFloat("normalmaps_smooth") == 0`. In this new code, `pow(2, 0)` is 1/1 (1). This does not look quite the same as 0.0078125

Comment (Zeno-) [inline src/shader.cpp]: `shaders_header += ftos(1 / powf(2, (128U << (u8)g_settings->getFloat("normalmaps_smooth"))));` might be correct (with sanity checking)

**Comment by Zeno-** [comment]
I'm not sure this is backwards compatible

**Comment by HybridDog** [comment]
>This does not look quite the same as 0.0078125

>and it changes the use of the normalmaps_smooth setting a bit

l think you should have the ability to use useless 1x1 generated normalmaps if you want.

**Comment by HybridDog** [comment]
>and it changes the use of the normalmaps_smooth setting a bit

>I'm not sure this is backwards compatible

l think it doesn't need to be backwards compatible, you would only need to update this small setting.

**Comment by ShadowNinja** [comment]
@HybridDog Why did you close this?

**Comment by HybridDog** [comment]
l thought it's low priority and only a small change, and so it's not really necessary.

**Comment by ShadowNinja** [comment]
It's not necessary, but that doesn't mean it's unwanted.

**Comment by nerzhul** [comment]
@ShadowNinja  :+1:

**Comment by HybridDog** [comment]
rebased.

**Pull request closed without merging**